ENH: Set Seeds as conntainer objects

blowekamp opened the pull request:
Best practices are to set member variables as object, and not repeat accessor methods in the interface of the owning object e.g. AddSeed, ClearSeed.

<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the ITK repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->

blowekamp commented:
@jhlegarreta  I was surprised to see `SetSeed[12]` marked as deprecated in some documentation changes made in #3866. d58bf56f 

These were the quick changes I made to have the seeds list behave like objects with setter/getters. Further investigation found other region growing filters with similar interfaces:
 * ConfidenceConnectedImageFilter
 * ConnectedThresholdImageFilter
 *  NeighborhoodConnectedImageFilter
 *  VectorConfidenceConnectedImageFilter
They all have AddSeed, ClearSeeds, methods. Some have the "SetSeed" method marked as deprecated in the doxygen. I didn't see any with a get method for the seed list.

These classes seem to use std::vector to hold the seed lists. Perhaps performance was a reason for this approach? Modern C++ could do something like `filter->SetSeedList(std::move(my_list))` where the data could be moved and not copied.

What can be done to improve the interface of these filters and make them more consistent?

jhlegarreta commented on Modules/Segmentation/RegionGrowing/include/itkIsolatedConnectedImageFilter.h:
```diff
@@ -124,15 +127,18 @@ class ITK_TEMPLATE_EXPORT IsolatedConnectedImageFilter : public ImageToImageFilt
    * This seed will be isolated from Seed1 (if possible).
    *
-   * \deprecated Please use AddSeed2.
+   * \deprecated Please use SetSeed2.
```
This should then say `SetSeeds2`.

hjmjohnson replied:
Fixed in 6ecbb73 — both \`AddSeed2\` and \`SetSeed2\` deprecation notices now point to \`SetSeeds2\` (matching the \`SetSeeds1\` pair). Rebased onto current \`main\` and resolved the \`@ITKStartGrouping\` conflict.

jhlegarreta reviewed (commented):
Really sorry @blowekamp, looks like I was confused by the docstring in the implementation file, which documented the method as being deprecated:
https://github.com/InsightSoftwareConsortium/ITK/commit/d58bf56fe2320fe3c7a451c96e3d8cd5e7c68358#diff-9daea1a5d3498baf38e16ef9fd06da89d2f898c4b599a7eb287a76fc3a720accL105
https://github.com/InsightSoftwareConsortium/ITK/commit/d58bf56fe2320fe3c7a451c96e3d8cd5e7c68358#diff-9daea1a5d3498baf38e16ef9fd06da89d2f898c4b599a7eb287a76fc3a720accL137

But after investigation, I now see that the documentation was not updated and the passage left inadvertently in
https://github.com/InsightSoftwareConsortium/ITK/commit/c005b54ffae285ec69bd219af39788fc2ae377d2

So please, go ahead and update the documentation.

If possible, can the methods be tested, please?

hjmjohnson commented:
Rebased onto current `main` (was ~7.5k commits behind / conflicting) and addressed @jhlegarreta's review: deprecation docstrings now point to `SetSeeds2`, and the test exercises the new `SetSeeds`/`GetSeeds` container accessors. Built `ITKRegionGrowingTestDriver` and ran both `itkIsolatedConnectedImageFilterTest` cases locally — green. Pushed to this branch (maintainer edit).

<details>
<summary>Details</summary>

- Conflict resolution: `main` wrapped `GetSeeds1/2` in `@ITKStartGrouping`/`@ITKEndGrouping`; kept the `itkSetMacro`/`itkGetConstReferenceMacro` block inside those markers.
- Test: round-trips the populated seed containers through `ClearSeeds` → `SetSeeds1/2` → `GetSeeds1/2`, leaving the filter state unchanged so the existing baseline comparison is unaffected.
- `pre-commit run --all-files` clean on the pushed tip.

</details>

<!--
provenance: claude-code gh-triage-pr session 2026-06-06
pushed_sha: 6ecbb73e88cc3266ca7056b30a8337bf9e5306d6 (force-with-lease over d5a4ea050b)
note: commit subject still has typo "conntainer"->"container"; left untouched to preserve authorship/message.
-->

hjmjohnson commented:
@blowekamp I've rebased your initial work and addressed the comment to add a test. I left this as a draft and kept you as the author.

hjmjohnson commented on Modules/Segmentation/RegionGrowing/include/itkIsolatedConnectedImageFilter.h:
```diff
@@ -125,15 +128,18 @@ class ITK_TEMPLATE_EXPORT IsolatedConnectedImageFilter : public ImageToImageFilt
+   *
+   * \deprecated Please use SetSeeds2.
+   */
   void
```
Should we be even more agressive here with ITK_FUTURE_LEGACY_REMOVE and warnings if used?

blowekamp replied:
I am working on another PR for the AddSeed/SetSeed methods.

hjmjohnson reviewed (approved):
I'm good with this as is, or with more aggressive deprecation warnings.

blowekamp committed:
ENH: Add SetSeeds/GetSeeds container interface to region growing filters

Add SeedsContainerType alias and itkSetMacro/itkGetConstReferenceMacro
for Seeds to ConfidenceConnectedImageFilter,
ConnectedThresholdImageFilter, NeighborhoodConnectedImageFilter, and
VectorConfidenceConnectedImageFilter. Removes manual virtual GetSeeds()
implementations in favour of the macro-generated accessors. Consistent
with IsolatedConnectedImageFilter.

hjmjohnson commented:
@blowekamp Greptile suggestions look like appropriate minor improvements.

blowekamp commented on Modules/Segmentation/RegionGrowing/include/itkIsolatedConnectedImageFilter.h:
```diff
@@ -144,10 +144,10 @@ class ITK_TEMPLATE_EXPORT IsolatedConnectedImageFilter : public ImageToImageFilt
   /** @ITKStartGrouping */
-  virtual const SeedsContainerType &
-  GetSeeds1() const;
-  virtual const SeedsContainerType &
-  GetSeeds2() const;
+  itkSetMacro(Seeds1, SeedsContainerType);
```
The suggested documentation does not add more than the name of the functions and data types. Access is at least a different word, perhaps in strict OO terminology a accessor is only a getter, and an access modifier (public, protected, private). Generally, the term access mean: "generally refers to the permission, ability, or right to approach, enter, make use of, or communicate with a person, place, or thing"

blowekamp committed:
ENH: Test SetSeeds/GetSeeds container interface in region growing filters

Add round-trip smoke tests for SetSeeds/GetSeeds to
ConfidenceConnectedImageFilter, ConnectedThresholdImageFilter,
NeighborhoodConnectedImageFilter, and
VectorConfidenceConnectedImageFilter tests. Verifies ClearSeeds empties
the container and SetSeeds/GetSeeds round-trip correctly.

hjmjohnson commented:
/azp run ITK.Linux

hjmjohnson commented:
/azp run ITK.Linux.Python

hjmjohnson commented:
/azp run ITK.macOS.Python

hjmjohnson commented on Modules/Segmentation/RegionGrowing/test/itkIsolatedConnectedImageFilterTest.cxx:
```diff
@@ -75,6 +75,29 @@ itkIsolatedConnectedImageFilterTest(int argc, char * argv[])
     filter->AddSeed2(seed1);
   }
 
+  // Exercise the seed-container accessors. Round-trip the populated
+  // containers through SetSeeds/GetSeeds, leaving the filter unchanged.
+  const FilterType::SeedsContainerType seeds1 = filter->GetSeeds1();
+  const FilterType::SeedsContainerType seeds2 = filter->GetSeeds2();
+
+  filter->ClearSeeds1();
+  filter->ClearSeeds2();
+  if (!filter->GetSeeds1().empty() || !filter->GetSeeds2().empty())
+  {
+    std::cerr << "Test failed!" << std::endl;
+    std::cerr << "Seed containers were not cleared by ClearSeeds." << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  filter->SetSeeds1(seeds1);
+  filter->SetSeeds2(seeds2);
+  if (filter->GetSeeds1() != seeds1 || filter->GetSeeds2() != seeds2)
+  {
+    std::cerr << "Test failed!" << std::endl;
+    std::cerr << "Seed containers retrieved by GetSeeds do not match those set by SetSeeds." << std::endl;
+    return EXIT_FAILURE;
+  }
+
   // The min and max values for a .png image
   constexpr FilterType::InputImagePixelType lower{ 0 };
   filter->SetLower(lower);
```
Follow up commit appears to have resolved this.
